File: 76-MinimumWindowSubstring/76-MinimumWindowSubstring.cpp
// Last updated: 7/27/2026, 10:15:24 PM
1class Solution {
2public:
3    string minWindow(string s, string t) {
4        int m = s.size();
5        int n = t.size();
6        if(n > m) return "";
7
8        int start = -1;
9        int l = 0;
10        int mini = INT_MAX;
11        int req = 0;
12        vector<int> v(256, 0);
13
14        for(int i = 0; i < n; i++) {
15            v[t[i]]++;
16            req++;
17        }
18
19        for(int i = 0; i < m; i++) {
20            if(v[s[i]] > 0) {
21                req--;
22                
23            }
24            v[s[i]]--;
25
26            while(req == 0) {
27                if(i - l + 1 < mini) {
28                    mini = i - l + 1;
29                    start = l;
30                }
31
32                v[s[l]]++;
33                if(v[s[l]] > 0) req++;
34                l++;
35            }
36
37        }
38        return start == -1 ? "" : s.substr(start, mini);
39    }
40};

File: 76-MinimumWindowSubstring/76-MinimumWindowSubstring.java
// Last updated: 8/8/2026, 6:57:08 PM
1class Solution {
2    public String minWindow(String s, String t) {
3        int m = s.length();
4        int n = t.length();
5        if(n > m) return "";
6        int[] arr = new int[256];
7        int mini = (int)1e5;
8        int l = 0;
9        int req = 0;
10        int start = -1;
11
12        for(int i = 0; i < n; i++) {
13            arr[t.charAt(i)]++;
14            req++;
15        }
16
17        for(int i = 0; i < m; i++) {
18            if(arr[s.charAt(i)] > 0) {
19                req--;
20            }
21            arr[s.charAt(i)]--;
22
23            while(req == 0) {
24                if((i - l + 1) < mini) {
25                    mini = i - l + 1;
26                    start = l;
27                }
28
29                arr[s.charAt(l)]++;
30                if(arr[s.charAt(l)] > 0) req++;
31                l++;
32            }
33        }
34        return start == -1 ? "" : s.substring(start, start + mini);
35    }
36}